features/268-next-cf/api.md

```diff
@@ -0,0 +1,625 @@
+
+To achieve client side routing in an SPA, all page routes should be served with `build/index.html`. On the other hand, js files, stylesheets, images and any static assets are put in the `build/asset` folder and would be served to requests path start with `/static`.
+
+A common `index.html` file may look like this
+
+```html
+<!doctype html>
+<html lang="en">
+  <head>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
+    <meta name="theme-color" content="#000000">
+    <link rel="stylesheet" href="/static/css/bootstrap.min.css">
+    <title>My App</title>
+  </head>
+  <body>
+    <noscript>
+      You need to enable JavaScript to run this app.
+    </noscript>
+    <div id="root"></div>
+    <script type="text/javascript" src="/static/js/bundle.js"></script>
+    <script type="text/javascript">
+      // The following is to demo calling the api server
+      var xhr = new XMLHttpRequest();
+      xhr.open("POST", "/api/greet", true);
+      xhr.setRequestHeader("Content-Type", "application/json");
+      xhr.onreadystatechange = function() { // Call a function when the state changes.
+          if (this.readyState === XMLHttpRequest.DONE && this.status === 200) {
+              // Request finished. Do processing here.
+          }
+      }
+      xhr.send("{}");
+    </script>
+  </body>
+</html>
+```
+
+## Server side rendering and api server
+
+```yaml
+cf:
+  ssr-server:
+    type: http-server
+    path: /
+    env: js
+    src: web
+  api-server:
+    type: http-service
+    path: /api
+    env: golang
+    src: api
+    secrets:
+      - DATABASE_URL
+
+static:
+  - src: asset
+    path: /static
+```
+
+# Security Concern
+
+## Restricted access to skygear gateway and gear pods in the same k8s cluster
+
+By default, all pods on the same k8s cluster will have a pod IP which is accessable by other pods in the same cluster. Since cloud function pods are in the same pods of skygear gateway and gear pods, cloud functions can call API to those pods directly.
+
+Skygear gateway is responsible for injecting tenant configuration and access key type to the request that is forwarded to the gears. And unless the gears verify the requests from the gateway, it would be unprotected, e.g. cloud function can act as another app by injecting tenant configuration in the header.
+
+We have two options:
+- Gears need to verify requests from gateway, maybe by having a shared secret
+- Applying network policy to restrict traffic between gear pods and other pods, https://kubernetes.io/docs/concepts/services-networking/network-policies/#isolated-and-non-isolated-pods
+
+# Features
+
+## Private functions / Micro-service discoverability
+
+- If a pod needs to be discovered by external network, services need to be setup
+- CF Gateway may block function request from outside if a function is marked as private
```

Comment:
I think auth hooks should be default private cloud function. Right?

Reply:
Yes.